packages/sdk/react-client/src/testing/withClientProvider.tsx

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { type Decorator } from '@storybook/react';
+import React, { useRef } from 'react';
+import { type FallbackProps, ErrorBoundary } from 'react-error-boundary';
+
+import { Trigger } from '@dxos/async';
+import { type Client } from '@dxos/client';
+import { type Space } from '@dxos/client/echo';
+import { performInvitation, TestBuilder } from '@dxos/client/testing';
+import { type MaybePromise } from '@dxos/util';
+
+import { ClientProvider, type ClientProviderProps } from '../client';
+
+type InitializeProps = {
+  createIdentity?: boolean;
+  createSpace?: boolean;
+  onSpaceCreated?: (client: Client, space: Space) => MaybePromise<void>;
+};
+
+/**
+ * Initialize client, identity and first space.
+ */
+const initialize = async (
+  client: Client,
+  {
+    createIdentity,
+    createSpace,
+    onSpaceCreated,
+    onInitialized,
+  }: InitializeProps & Pick<WithClientProviderProps, 'onInitialized'>,
+): Promise<Space | undefined> => {
+  if (createIdentity) {
+    await client.halo.createIdentity();
+  }
+
+  let space: Space | undefined;
+  if (createSpace) {
+    space = await client.spaces.create({ name: 'Test Space' });
+    await onSpaceCreated?.(client, space);
+  }
+
+  await onInitialized?.(client);
+  return space;
+};
+
+export type WithClientProviderProps = InitializeProps & ClientProviderProps;
+
+/**
+ * Decorator that provides the client context.
+ */
+export const withClientProvider = ({
+  createIdentity,
+  createSpace,
+  onSpaceCreated,
+  onInitialized,
+  ...props
+}: WithClientProviderProps = {}): Decorator => {
+  const handleInitialized = async (client: Client) => {
+    await initialize(client, { createIdentity, createSpace, onSpaceCreated, onInitialized });
+  };
+
+  return (Story) => (
+    <ErrorBoundary FallbackComponent={ErrorFallback}>
+      <ClientProvider onInitialized={handleInitialized} {...props}>
+        <Story />
+      </ClientProvider>
+    </ErrorBoundary>
+  );
+};
+
+// TODO(burdon): Controls.
+// TODO(burdon): Callback once all invitations have completed.
+// TODO(burdon): Control delay/jitter for creation of other clients.
+export type WithMultiClientProviderProps = InitializeProps & ClientProviderProps & { numClients?: number };
+
+/**
+ * Decorator that creates a scaffold for multiple clients.
+ * Orchestrates invitations between a randomly selected host and the remaining clients.
+ */
+export const withMultiClientProvider = ({
+  numClients = 2,
```

Comment:
```suggestion
  count = 2,
```

Reply:
i go back and forth. count is good, until later when you have many options so i'm currently in my more verbose phase. thinking.